### PR TITLE
feat(mode): switch to `ModeChanged` for more detailed mode information

### PR DESCRIPTION
**Pull request by theol0403**
This gives us information for when the mode switches between kinds of visual modes. 

Based on https://github.com/vscode-neovim/vscode-neovim/pull/1254

### CODE REVIEW

**Comment by theol0403** [comment]
My only concern with this is:
- am I parsing the modes (especially visual-block) correctly, or is there a more idiomatic way of doing this
- is there any special vim-mode that we would benefit from parsing correctly
- `mode_change` always fired on buffer init, wheras `ModeChanged` does not. Is `BufAdd` the best autocommand to ensure the mode is set on init? 

This is used for my upcoming visual selection PR, which needs to know the difference between cursor block etc.

File: src/mode_manager.ts [hunk 1]
```diff
@@ -3,17 +3,45 @@ import { EventEmitter } from "events";
...
+                return "insert";
+            case "v":
+                return "visual";
+            case "n":
+            default:
+                return "normal";
+        }
+    }
+    // visual mode name
+    public get visual(): "single" | "line" | "block" {
```
Comment (justinmk) [inline src/mode_manager.ts:31]: the usual names are charwise, linewise, block. `:help charwise`.

```suggestion
    public get visual(): "char" | "line" | "block" {
```

File: src/mode_manager.ts [hunk 2]
```diff
@@ -3,17 +3,45 @@ import { EventEmitter } from "events";
...
+// a representation of the current mode. can be read in different ways using accessors. underlying type is full name as returned by `:help mode()`
+class Mode {
+    public constructor(public full: string = "") {}
+    // mode 1-char code: n, v, V, i, s, ...
+    // converts ^v into v
+    public get char(): string {
+        return this.full.charCodeAt(0) == 22 ? "v" : this.full.charAt(0);
+    }
+    // mode full name
+    public get long(): "insert" | "visual" | "normal" {
```
Comment (justinmk) [inline src/mode_manager.ts:19]: how does this differ from `this.full`? could be a confusing interface.

```suggestion
    public get name(): "insert" | "visual" | "normal" {
```

**Comment by justinmk** [comment]
> am I parsing the modes (especially visual-block) correctly, or is there a more idiomatic way of doing this

This is the normal way.

File: src/mode_manager.ts (hunk 2 above)
Comment (theol0403) [reply src/mode_manager.ts:19]: Thanks for the catch. I renamed this.full to `raw`, which is the shortcode returned by `mode()`.
`long` is the mode name. 

It was confusing indeed, this was why CI is failing lol.

Comment (theol0403) [reply src/mode_manager.ts:19]: On second thought, I prefer your suggestion for `name`.

**Review by justinmk — APPROVED** [review]

**Pull request merged**